pyscf/pbc/symm/group.py

```diff
@@ -64,244 +132,331 @@ def matrix(self, a):
+        if self is other:
+            return self
+        hi = list(self.hash_table.keys())
+        hj = list(other.hash_table.keys())
+        hij = np.union1d(hi, hj)
+        return self.__class__(hij, from_hash=True)
+
+    def issubset(self, other):
+        if type(self) is not type(other):
+            raise TypeError(f'{self} and {other} must be the same type')
+        if self is other:
+            return True
+        hi = list(self.hash_table.keys())
+        hj = list(other.hash_table.keys())
+        return set(hi).issubset(set(hj))
+
     @property
-    def group_name(self):
-        return geom.get_crystal_class(None, self.elements)[0]
+    def order(self):
+        if self._order is None:
+            self._order = len(self.elements)
+        return self._order
 
-    def lookup_table(self):
+    @order.setter
+    def order(self, n):
+        self._order = n
+
+    @property
+    def hash_table(self):
         '''
-        Mappings between elements' hash values and their indices.
+        Hash table for group elements: {hash : index}.
         '''
-        return {hash(g) : i for i, g in enumerate(self.elements)}
+        if self._hash_table is None:
+            self._hash_table = {hash(g) : i for i, g in enumerate(self.elements)}
+        return self._hash_table
+
+    @hash_table.setter
+    def hash_table(self, table):
+        self._hash_table = table
 
+    @property
     def inverse_table(self):
         '''
-        Inverse of the elements. Returns the indices.
+        Table for inverse of the group elements.
+
+        Return : (n,) array of ints
+            The indices of elements.
         '''
-        n = len(self)
-        inv_table = np.zeros((n,), dtype=np.int32)
-        lookup = self.lookup_table()
-        for i, g in enumerate(self.elements):
-            inv_table[i] = lookup[hash(g.inv())]
-        return inv_table
+        if self._inverse_table is None:
+            _table = [self.hash_table[hash(g.inv())] for g in self.elements]
+            self._inverse_table = np.asarray(_table)
+        return self._inverse_table
 
+    @inverse_table.setter
+    def inverse_table(self, table):
+        self._inverse_table = table
+
+    @property
     def multiplication_table(self):
         '''
-        Multiplication table of the group. Returns the indices of elements.
+        Multiplication table of the group.
+
+        Return : (n, n) array of ints
+             The indices of elements.
         '''
-        n = len(self)
-        prod_table = np.zeros((n,n), dtype=np.int32)
-        lookup = self.lookup_table()
-        for i, g in enumerate(self.elements):
-            for j, h in enumerate(self.elements):
-                prod = g @ h
-                prod_table[i,j] = lookup[hash(prod)]
-        return prod_table
+        if self._multiplication_table is None:
+            prod = self.elements[:,None] * self.elements[None,:]
+            _table = [self.hash_table[hash(gh)] for gh in prod.flatten()]
```

Comment:
Multiplication table requires elements to be unique. Is it possible collision in hash table?

Reply:
Added a sanity check function to make sure the elements form a group.